app/webpacker/lib/hooks/useStoredReducer.js

```diff
@@ -0,0 +1,33 @@
+import { useReducer } from 'react';
+
+/**
+ * This functions like the useReducer hook, but it fetches the state stored in
+ * local storage, via the given key, on subsequent uses.
+ *
+ * Do NOT call this twice with the same key - updating one of such a pair
+ * will not update the other's state.
+ */
+export default function useStoredReducer(reducer, initialState, key) {
+  let storedState;
+  try {
+    storedState = JSON.parse(localStorage.getItem(key));
+  } catch {
+    storedState = null;
+  }
+
+  function augmentedReducer(state, action) {
+    const newState = reducer(state, action);
+    localStorage.setItem(key, JSON.stringify(newState));
+    return newState;
+  }
```

Comment:
Would it be worth wrapping this in a `useCallback`?

Reply:
I don't think there's any benefit. It's not being used in any dependency arrays that I'm aware of. Reducers need to be pure (not depend on any state), but it doesn't matter if the reference is a new function every time.